File: LeetCode/longest_common_prefix.go
// O(S), O(1), S is the sum of all characters in all strings
func longestCommonPrefix(strs []string) string {
    pickedStr := strs[0]
    lenPickedStr := len(pickedStr)
    for i := 1; i < len(strs); i++ {
        if lenPickedStr == 0 {
            break
        }
        
        str := strs[i]
        lenStr := len(str)
        if lenStr == 0 {
            return ""
        }
        
        
        for j := 0; j < lenStr && j < lenPickedStr; j++ {
            if str[j] != pickedStr[j] {
                pickedStr = str[:j]
                lenPickedStr = j
                break
            }
            
            if j + 1 == lenStr {
                pickedStr = str[:j + 1]
                lenPickedStr = j + 1
            }
        }
    }
    
    return pickedStr
}

File: LeetCode/longest_common_prefix.py
class Solution:
    # O(S), O(1), S is the sum of all characters in all strings
    def longestCommonPrefix(self, strs: List[str]) -> str:
        len_strs = len(strs)
        if len_strs == 0: return ''
        if len_strs == 1: return strs[0]
        
        ref = strs[0]
        for i in range(1, len_strs):
            min_len = min(len(ref), len(strs[i]))
            
            tmps = ''
            for j in range(min_len):
                if ref[j] == strs[i][j]: tmps += ref[j]
                else: break
                        
            ref = tmps
            if ref == '': return ''
        
        return ref
        